src/main/java/com/googleintern/wfm/ruleengine/action/RuleValidation.java

```diff
@@ -0,0 +1,287 @@
+    for (Long skillId : skillIds) {
+      skillIdBuilder.append(skillIdBuilder.length() == 1 ? "" : Separator.COMMA.symbol);
+      skillIdBuilder.append(skillId);
+    }
+    for (Long roleSkillId : roleSkillIds) {
+      skillIdBuilder.append(skillIdBuilder.length() == 1 ? "" : Separator.COMMA.symbol);
+      skillIdBuilder.append(roleSkillId);
+    }
+    skillIdBuilder.append(Separator.SQUARE_BRACKET_RIGHT.symbol);
+    return skillIdBuilder.toString();
+  }
+
+  private static String convertPoolAssignmentsToString(Set<PoolAssignmentModel> poolAssignments) {
+    StringBuilder poolAssignmentsStringBuilder = new StringBuilder();
+    poolAssignmentsStringBuilder.append(Separator.SQUARE_BRACKET_LEFT.symbol);
+    for (PoolAssignmentModel poolAssignment : poolAssignments) {
+      poolAssignmentsStringBuilder.append(
+          poolAssignmentsStringBuilder.length() == 1 ? "" : Separator.COMMA.symbol);
+      poolAssignmentsStringBuilder.append(
+          CASE_POOL_ID_PREFIX
+              + poolAssignment.casePoolId()
+              + PERMISSION_SET_ID_PREFIX
+              + poolAssignment.permissionSetId()
+              + Separator.DOUBLE_QUOTATION_MARK.symbol
+              + Separator.CURLY_BRACKET_RIGHT.symbol);
+    }
+    poolAssignmentsStringBuilder.append(Separator.SQUARE_BRACKET_RIGHT.symbol);
+    return poolAssignmentsStringBuilder.toString();
+  }
+
+  private static ImmutableSetMultimap<UserPoolAssignmentModel, PoolAssignmentModel>
+      assignPermissionsBasedOnGeneratedRules(
+          List<UserPoolAssignmentModel> expectedUserPoolAssignments) {
+    ImmutableSetMultimap.Builder<UserPoolAssignmentModel, PoolAssignmentModel>
+        filtersByUserPoolAssignmentBuilder = ImmutableSetMultimap.builder();
+    for (UserPoolAssignmentModel user : expectedUserPoolAssignments) {
+      filtersByUserPoolAssignmentBuilder.putAll(
+          user,
+          assignPermissions(
+              user,
+              ImmutableSet.copyOf(
+                  CasePoolIdAndPermissionIdGroupingUtil.convertSkillIdRoleIdToFilter(user))));
+    }
+    return filtersByUserPoolAssignmentBuilder.build();
+  }
+
+  private static ImmutableSet<PoolAssignmentModel> assignPermissions(
+      UserPoolAssignmentModel user, ImmutableSet<FilterModel> userFilters) {
+    return rules.stream()
+        .filter(
+            rule ->
+                (decideToAssignPermissions(
+                    user.workforceId(), user.workgroupId(), userFilters, rule)))
+        .flatMap(
+            rule ->
+                rule.permissionSetIds().stream()
+                    .map(
+                        permissionSetId ->
+                            PoolAssignmentModel.builder()
+                                .setCasePoolId(rule.casePoolId())
+                                .setPermissionSetId(permissionSetId)
+                                .build()))
+        .collect(ImmutableSet.toImmutableSet());
+  }
+
+  private static boolean decideToAssignPermissions(
```

Comment:
For methods that return a boolean, the name should be descriptive. 
```
isEmpty()      true -> it is empty     false -> it is not empty
shouldAssignPermission()    true -> yes we should.     false -> no we should not
```

decideToAssignPermissions ?